### DISCUSSION TITLE
How to get a raw string for a specified AST node?

### DISCUSSION THREAD

**Discussion by liuliangsir** [babel/babel#12286] [Q&A] [answered]
As I mentioned earlier, I would like to get a raw string for `TemplateLiteral` and `(Arrow)FunctionExpression` when walking the whole AST. That is to say, I would like to get a expected data structure during traverse. The followings are a bunch of code snippets. Is there any approach to easily get a raw string for a specified AST node? Is it possible to add a method(toRawString) on AST node to get a raw string?

Convert
```js
await ac.sysDialogOn({
  content:`${1+2+2}a`,
  id: 2330,
  hasRoleName: true,
  hasBg: false,
  hasRoleAvatar: true,
  roleName: `角色名`,
  roleAvatarResId: '$262638',
});
```
to

```js
{
  "type": "avgApi",
  "payload": {
    "content": "${1+2+2}a",
    "id": 2330,
    "hasRoleName": false,
    "hasBg": false,
    "hasRoleAvatar": true,
    "roleName": "角色wwwww名",
    "roleAvatarResId": "$262638"
  }
}
```

**Comment by liuliangsir** [comment, marked as answer]
@nicolo-ribaudo cc

**Reply by nicolo-ribaudo** [reply]
- If you want to get the original source corresponding to a node, you can use the location data:
  ```js
  function getNodeSource(fileContents, node) {
    return fileContents.slice(node.start, node.end);
  }
  ```

- If you want to partially evaluate a node, you can try [`path.evaluate()`](https://github.com/babel/babel/blob/b9407d766062118660faab66bf7ce0d9ef1bcf74/packages/babel-traverse/src/path/evaluation.js#L407-L421).

**Comment by liuliangsir** [comment]
@nicolo-ribaudo I am using `@babel/generator`. But I think your answer is a better choice.
